extensions/azure/fcc-node-directory-cosmos/src/main/java/org/eclipse/dataspaceconnector/catalog/node/store/cosmos/CosmosFederatedCacheNodeStore.java

```diff
@@ -0,0 +1,113 @@
+import com.azure.cosmos.models.CosmosItemRequestOptions;
+import com.azure.cosmos.models.CosmosItemResponse;
+import com.azure.cosmos.models.CosmosQueryRequestOptions;
+import com.azure.cosmos.models.PartitionKey;
+import net.jodah.failsafe.RetryPolicy;
+import org.eclipse.dataspaceconnector.catalog.node.store.cosmos.model.FederatedCacheNodeDocument;
+import org.eclipse.dataspaceconnector.catalog.spi.FederatedCacheNode;
+import org.eclipse.dataspaceconnector.catalog.spi.FederatedCacheNodeDirectory;
+import org.eclipse.dataspaceconnector.spi.EdcException;
+import org.eclipse.dataspaceconnector.spi.types.TypeManager;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static net.jodah.failsafe.Failsafe.with;
+
+public class CosmosFederatedCacheNodeStore implements FederatedCacheNodeDirectory {
+
+    private final CosmosContainer container;
+    private final CosmosQueryRequestOptions tracingOptions;
+    private final TypeManager typeManager;
+    private final String partitionKey;
+    private final RetryPolicy<Object> retryPolicy;
+
+    /**
+     * Creates a new instance of the CosmosDB-based federated cache node store.
+     *
+     * @param container    The CosmosDB-container that'll hold the federated cache nodes.
+     * @param typeManager  The {@link TypeManager} that's used for serialization and deserialization
+     * @param partitionKey A Partition Key that CosmosDB uses for r/w distribution. Contrary to what CosmosDB suggests, this
+     *                     key should be the same for all local (=clustered) connectors, otherwise queries in stored procedures might
+     *                     produce incomplete results.
+     */
+    public CosmosFederatedCacheNodeStore(CosmosContainer container, TypeManager typeManager, String partitionKey) {
+
+        this.container = container;
+        this.typeManager = typeManager;
+        this.partitionKey = partitionKey;
+        tracingOptions = new CosmosQueryRequestOptions();
+        tracingOptions.setQueryMetricsEnabled(true);
+        retryPolicy = new RetryPolicy<>()
+                .handle(RequestRateTooLargeException.class)
+                .withMaxRetries(1).withDelay(Duration.ofSeconds(5));
+    }
+
+    @Override
+    public List<FederatedCacheNode> getAll() {
+        var query = "SELECT * FROM FederatedCatalogueNodeDocument";
+
+        try {
+            var response = with(retryPolicy).get(() -> container.queryItems(query, tracingOptions, Object.class));
+            return response.stream()
+                    .map(this::convertObject)
+                    .map(FederatedCacheNodeDocument::getWrappedInstance)
+                    .collect(Collectors.toList());
+        } catch (CosmosException ex) {
+            throw new EdcException(ex);
+        }
+    }
+
+    @Override
+    public Stream<FederatedCacheNode> getAllAsync() {
+        return null;
```

Comment:
ideally, this should just return `Stream.of(getAll())`. the async variant is just there in case the backend supports it.